### PR TITLE
test(multichain): fusdc "advance and settle" as a macro

### PR DESCRIPTION
**Pull request by 0xpatrickdev**
refs: #10597

## Description
- tests additional FUSDC happy paths with agoric and noble EUDs
- "oracles accept" test conditionally accepts invitations

### Security Considerations
None, only tests

### Scaling Considerations
None, ~60s hit to CI (each advance + settle is about 30s)

### Documentation Considerations
None

### Testing Considerations
PR is only tests

### Upgrade Considerations
None

### CODE REVIEW

File: multichain-testing/test/fast-usdc/fast-usdc.test.ts [hunk 1]
```diff
@@ -34,7 +34,7 @@ const accounts = [...keys(oracleMnemonics), 'lp'];
 const contractName = 'fastUsdc';
 const contractBuilder =
   '../packages/builders/scripts/fast-usdc/init-fast-usdc.js';
-const LP_DEPOSIT_AMOUNT = 10_000_000n;
+const LP_DEPOSIT_AMOUNT = 8_000n * 10n ** 6n;
```
Comment (turadg) [inline multichain-testing/test/fast-usdc/fast-usdc.test.ts:37]: we have the 10^6 as a constant somewhere but this is clear enough. better than the literal

File: multichain-testing/test/fast-usdc/fast-usdc.test.ts [hunk 2]
```diff
@@ -103,21 +103,44 @@ const toOracleOfferId = (idx: number) => `oracle${idx + 1}-accept`;
...
 test.serial('oracles accept', async t => {
   const { oracleWds, retryUntilCondition, vstorageClient, wallets } = t.context;
+  const brands = await vstorageClient.queryData('published.agoricNames.brand');
+  const { Invitation } = Object.fromEntries(brands);
 
-  const instances = await vstorageClient.queryData(
-    'published.agoricNames.instance',
-  );
-  const instance = fromEntries(instances)[contractName];
+  // TODO export/import INVITATION_MAKERS_DESC
```
Comment (turadg) [inline multichain-testing/test/fast-usdc/fast-usdc.test.ts]: should not be imported. the test should be declaring what it expects the value to be, not just that it matches something else. 

consider if someone changed the value in the other module. the test should fail and have to be updated to make clear that the API  changed.

File: multichain-testing/test/fast-usdc/fast-usdc.test.ts [hunk 3]
```diff
@@ -103,21 +103,44 @@ const toOracleOfferId = (idx: number) => `oracle${idx + 1}-accept`;
...
+  const ORACLE_INV_DESC = 'oracle operator invitation';
```
Comment (turadg) [inline multichain-testing/test/fast-usdc/fast-usdc.test.ts]: for the reason above I think this shouldn't have ALL_CAPS casing. defining the value at all is just a convenience in this function to not have to repeat it.
```

```suggestion
  const description = 'oracle operator invitation';
```

**Review by turadg — APPROVED** [review]

File: multichain-testing/test/fast-usdc/fast-usdc.test.ts [hunk 4]
```diff
@@ -103,21 +103,44 @@ const toOracleOfferId = (idx: number) => `oracle${idx + 1}-accept`;
...
+  let hasAccepted = false;
+  for (const name of keys(oracleMnemonics)) {
+    const { offerToUsedInvitation, purses } = await vstorageClient.queryData(
+      `published.wallet.${wallets[name]}.current`,
+    );
+    const { value: invitations } = balancesFromPurses(purses)[Invitation];
+    const hasInvitation = invitations.some(
+      x => x.description === ORACLE_INV_DESC,
+    );
+    const usedInvitation = offerToUsedInvitation?.[0]?.[0] === `${name}-accept`;
```
Comment (dckc) [inline multichain-testing/test/fast-usdc/fast-usdc.test.ts:119]:
```suggestion
    const usedInvitation = offerToUsedInvitation.some(([id, _v]) => id === `${name}-accept`);
```

File: multichain-testing/test/fast-usdc/fast-usdc.test.ts [hunk 5]
```diff
@@ -103,21 +103,41 @@ const toOracleOfferId = (idx: number) => `oracle${idx + 1}-accept`;
...
 
-  const instances = await vstorageClient.queryData(
-    'published.agoricNames.instance',
-  );
-  const instance = fromEntries(instances)[contractName];
+  const description = 'oracle operator invitation';
+
+  // ensure we have an unused (or used) oracle invitation in each purse
+  let hasAccepted = false;
+  for (const name of keys(oracleMnemonics)) {
```
Comment (dckc) [inline multichain-testing/test/fast-usdc/fast-usdc.test.ts:113]: There's a lot going on in this loop.

If we reify the oracle operators as objects somewhat like [makeOracleOperator](https://github.com/Agoric/agoric-sdk/blob/d35659b6ab227f11f952d864d38add58dd464825/packages/fast-usdc/test/fast-usdc.contract.test.ts#L208) or [makeAgoricWatcher](https://github.com/Agoric/agoric-sdk/blob/74d293b9fae15081b92e3876beba7ed6278e4b39/packages/fast-usdc/test/quickSend-tx.test.ts#L151), we can make this read much more straightforwardly:

```suggestion
  for (const op of oracleOperators) {
    const status = await oracle.invitationCheck(); // returns the invitation or `used` if used;
    t.log({ name: op.name, status });
    t.truthy(status);
  }
```

File: multichain-testing/test/fast-usdc/fast-usdc.test.ts [hunk 6]
```diff
@@ -103,21 +103,41 @@ const toOracleOfferId = (idx: number) => `oracle${idx + 1}-accept`;
...
+      `published.wallet.${wallets[name]}.current`,
+    );
+    const { value: invitations } = balancesFromPurses(purses)[Invitation];
+    const hasInvitation = invitations.some(x => x.description === description);
+    const usedInvitation = offerToUsedInvitation?.[0]?.[0] === `${name}-accept`;
+    t.log({ name, hasInvitation, usedInvitation });
+    t.true(hasInvitation || usedInvitation, 'has or accepted invitation');
+    if (usedInvitation) hasAccepted = true;
+  }
+  // if the oracles have already accepted, skip the rest of the test this is
```
Comment (dckc) [inline multichain-testing/test/fast-usdc/fast-usdc.test.ts:124]:
```suggestion
  // if any of the oracles have already accepted, skip the rest of the test this is
```

right?

File: multichain-testing/test/fast-usdc/fast-usdc.test.ts [hunk 7]
```diff
@@ -187,129 +207,136 @@ test.serial('lp deposits', async t => {
...
-  const eudWallet = await createWallet(useChain('osmosis').chain.bech32_prefix);
-  const EUD = (await eudWallet.getAccounts())[0].address;
-
-  // parameterize agoric address
-  const { settlementAccount } = await vstorageClient.queryData(
-    `published.${contractName}`,
-  );
-  t.log('settlementAccount address', settlementAccount);
+const advanceAndSettleScenario = test.macro({
+  title: (_, mintAmt: bigint, eudChain: string) =>
```
Comment (dckc) [inline multichain-testing/test/fast-usdc/fast-usdc.test.ts:211]: `eudChain` is kinda... tortured

```suggestion
  title: (_, mintAmt: bigint, destChain: string) =>
```

File: multichain-testing/test/fast-usdc/fast-usdc.test.ts [hunk 8]
```diff
@@ -187,129 +207,136 @@ test.serial('lp deposits', async t => {
...
+      nobleAgoricChannelId,
+      oracleWds,
+      retryUntilCondition,
+      useChain,
+      usdcOnOsmosis,
+      vstorageClient,
+    } = t.context;
+
+    // EUD wallet on the specified chain
+    const eudWallet = await createWallet(
```
Comment (dckc) [inline multichain-testing/test/fast-usdc/fast-usdc.test.ts:225]: likewise...

```suggestion
    const destWallet = await createWallet(
```

Comment (dckc) [reply multichain-testing/test/fast-usdc/fast-usdc.test.ts:225]: I'd rather avoid using the ambient random source inside `createWallet`. I don't suppose we have a widget where you pass `crypto.randomBytes(plenty)` and it makes a mnemonic, do we? Or can we pass in a private key?

I can imagine this isn't worthwhile just now, so... adding it to the list:

  - #2160

File: multichain-testing/test/fast-usdc/fast-usdc.test.ts [hunk 9]
```diff
@@ -187,129 +207,136 @@ test.serial('lp deposits', async t => {
...
-    blockNumber: 21037663n,
-    txHash: `0xc81bc6105b60a234c7c50ac17816ebcd5561d366df8bf3be59ff3875527617${makeRandomDigits(makeRandomNumber(), 2n)}`,
-    tx: {
-      amount: mintAmount,
-      forwardingAddress: userForwardingAddr,
-    },
-    aux: {
-      forwardingChannel: nobleAgoricChannelId,
+    // register forwarding address on noble
+    const txRes = nobleTools.registerForwardingAcct(
```
Comment (dckc) [inline multichain-testing/test/fast-usdc/fast-usdc.test.ts:241]: again, lots going on; my tiny brain wants the roles (user, noble chain, ...) reified as objects.

File: multichain-testing/test/fast-usdc/fast-usdc.test.ts [hunk 10]
```diff
@@ -187,129 +207,136 @@ test.serial('lp deposits', async t => {
...
+        recipientAddress,
+      },
+      chainId: 42161,
+    });
+
+    console.log('User initiates evm mint:', evidence.txHash);
+
+    // submit evidences
+    await Promise.all(
+      oracleWds.map(makeDoOffer).map((doOffer, i) =>
```
Comment (dckc) [inline multichain-testing/test/fast-usdc/fast-usdc.test.ts:274]: would love to see...

```suggestion
      oracleOperators.map(op => op.submit(evidence));
```

**Review by dckc — COMMENTED** [review]
I'd like to talk over a suggestion to factor out roles / participants with `makeOracleOperator(...)` and such.

File: multichain-testing/test/fast-usdc/fast-usdc.test.ts [hunk 11]
```diff
@@ -103,21 +103,41 @@ const toOracleOfferId = (idx: number) => `oracle${idx + 1}-accept`;
...
-  const instance = fromEntries(instances)[contractName];
+  const description = 'oracle operator invitation';
+
+  // ensure we have an unused (or used) oracle invitation in each purse
+  let hasAccepted = false;
+  for (const name of keys(oracleMnemonics)) {
+    const { offerToUsedInvitation, purses } = await vstorageClient.queryData(
+      `published.wallet.${wallets[name]}.current`,
+    );
+    const { value: invitations } = balancesFromPurses(purses)[Invitation];
```
Comment (dckc) [inline multichain-testing/test/fast-usdc/fast-usdc.test.ts:117]: Does that work?

`balancesFromPurses` returns a map, right? and `Invitation` is a `Brand`

File: multichain-testing/test/fast-usdc/fast-usdc.test.ts [hunk 12]
```diff
@@ -101,41 +108,91 @@ test.after(async t => {
...
       ),
-    );
+    ),
+  );
+
+  const checks = await Promise.all(
+    operators.map(async op => {
+      const { invitation, usedInvitation } = await op.checkInvitation();
+      t.log({ name: op.getName(), invitation, usedInvitation });
+      t.true(invitation || usedInvitation, 'has or accepted invitation');
```
Comment (samsiegart) [inline multichain-testing/test/fast-usdc/fast-usdc.test.ts]: Shouldn't this `map` function return `usedInvitation` for the logic below to work? If it doesn't return anything then `checks.some(Boolean)` would always be false right?

**Comment by dckc** [comment]
I haven't gotten set up to test this locally; using ci to test it is pretty slow.

Let's roll back to before @0xpatrickdev handed this to me and see if it's green. If so, I can do my refactor in a follow-on.

**Comment by dckc** [comment]
## **No quorum**: flake in z:acceptance governance?

[flake](https://github.com/Agoric/agoric-sdk/actions/runs/12345315961/job/34449230334?pr=10692#step:10:8388)?

```
[z:acceptance] Running test.sh.

ACCEPTANCE TESTING governance
...

  ✘ [fail]: economic committee can make governance proposal and vote on it
...
        reason: 'No quorum',
...
Error: Command failed: docker run -e SLOGFILE -v "$SLOGFILE:/home/runner/test.slog" --rm ghcr.io/agoric/agoric-3-proposals:test-acceptance
```
Mon, 16 Dec 2024 04:10:39 GMT

**Comment by dckc** [comment]
## psm acceptance flake?

in `a3p-integration/proposals/z:acceptance/psm.test.js`

https://github.com/Agoric/agoric-sdk/actions/runs/12363738833/job/34505666038?pr=10692#step:10:8349

**Pull request merged**